File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/text/AgentFrontmatterRule.java
package com.ririnto.sinon.harness.rules.text;

import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import com.ririnto.sinon.harness.rules.HarnessCheckRule;
import com.ririnto.sinon.harness.core.RuleContext;
import com.ririnto.sinon.harness.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Rule that requires agent files to have proper frontmatter.
 */
public enum AgentFrontmatterRule implements HarnessCheckRule {
    INSTANCE;

    private static final String CATEGORY = "agentFrontmatter";

    @Override
    public String category() {
        return "agentFrontmatter";
    }

    @Override
    public boolean applies(RuleContext ctx) {
        return ctx.manifest().isEnabled(CATEGORY);
    }

    @Override
    public Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException {
        final Path root = ctx.root();
        final JsonNode manifest = ctx.manifest().raw();
        final Path directory = root.resolve(".claude/agents");
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        if (!HarnessCheckHelper.isSafeDirectory(root, directory)) {
            return List.of(Finding.of(severity, CATEGORY, ".claude/agents must contain at least one .md agent"));
        }
        final List<Path> files;
        try (final Stream<Path> stream = Files.list(directory)) {
            files = stream
                    .filter(f -> !Files.isSymbolicLink(f))
                    .filter(f -> f.getFileName().toString().endsWith(".md"))
                    .toList();
        } catch (IOException e) {
            return List.of();
        }
        if (files.isEmpty()) {
            return List.of(Finding.of(severity, CATEGORY, ".claude/agents must contain at least one .md agent"));
        }
        final Pattern namePattern = Pattern.compile("(?m)^name:\\s*[-a-z0-9]+\\s*$");
        final Pattern descPattern = Pattern.compile("(?m)^description:\\s*.+$");
        return files.stream()
                .flatMap(f -> {
                    try {
                        return validateAgentFile(root, f, namePattern, descPattern, severity).stream();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }

    private List<Finding> validateAgentFile(Path root, Path file, Pattern namePattern, Pattern descPattern, String severity) throws MojoExecutionException {
        final String text = HarnessCheckHelper.readFile(root, file);
        final String relative = root.relativize(file).toString();
        return Stream.<Stream<Finding>>of(
                text.startsWith("---") ? Stream.<Finding>empty() : Stream.of(Finding.of(severity, CATEGORY, "agent missing frontmatter: " + relative)),
                namePattern.matcher(text).find() ? Stream.<Finding>empty() : Stream.of(Finding.of(severity, CATEGORY, "agent missing name: " + relative)),
                descPattern.matcher(text).find() ? Stream.<Finding>empty() : Stream.of(Finding.of(severity, CATEGORY, "agent missing description: " + relative))
        ).flatMap(s -> s).toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/text/AgentFrontmatterRule.kt
package com.ririnto.sinon.harness.rules.text

import com.ririnto.sinon.harness.core.JsonAccess
import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.rules.HarnessCheckRule
import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import kotlinx.serialization.json.jsonObject
import org.commonmark.ext.front.matter.YamlFrontMatterExtension
import org.commonmark.ext.front.matter.YamlFrontMatterVisitor
import org.commonmark.parser.Parser
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.extension
import kotlin.io.path.isDirectory
import kotlin.io.path.isRegularFile
import kotlin.io.path.listDirectoryEntries
import kotlin.io.path.readText
import kotlin.io.path.relativeTo

/**
 * Rule that requires agents to have proper frontmatter.
 *
 * @category agentFrontmatter
 */
object AgentFrontmatterRule : HarnessCheckRule() {
    /**
     * Category key.
     */
    override val category: String = "agentFrontmatter"

    private val namePattern: Regex = "[-a-z0-9]+".toRegex()

    private val markdownParser: Parser =
        Parser.builder()
            .extensions(listOf(YamlFrontMatterExtension.create()))
            .build()

    override fun validate(ctx: RuleContext): Collection<Finding> = buildList {
        val catObj = ctx.manifest.categoryObject(category)
        val parametersObj = catObj?.get("parameters")?.jsonObject
        if (catObj != null && parametersObj != null) {
            val severity = ctx.manifest.severityOf(category)
            val dirPath = ctx.root / JsonAccess.stringFromObject(parametersObj, "directory")
            when {
                !dirPath.isDirectory() -> add(
                    Finding(
                        severity,
                        category,
                        ctx.manifest.stringValue(category, "missingDirectory").takeIf { message ->
                            message.isNotEmpty()
                        } ?: ".claude/agents must contain at least one .md agent",
                    ),
                )
                else -> {
                    val files =
                        dirPath
                            .listDirectoryEntries()
                            .filter { entry -> entry.isRegularFile() }
                            .filter { entry -> entry.extension == "md" }
                    if (files.isEmpty()) {
                        add(
                            Finding(
                                severity,
                                category,
                                ctx.manifest.stringValue(category, "missingAgent").takeIf { message ->
                                    message.isNotEmpty()
                                } ?: ".claude/agents must contain at least one .md agent",
                            ),
                        )
                    }
                    files.forEach { file ->
                        val text = file.readText()
                        when {
                            !text.startsWith("---") -> add(
                                Finding(
                                    severity,
                                    category,
                                    ctx.manifest.stringValue(category, "missingFrontmatter").takeIf { message ->
                                        message.isNotEmpty()
                                    } ?: "agent missing frontmatter: ${file.relativeTo(ctx.root)}",
                                ),
                            )
                            else -> {
                                val frontmatter = extractFrontmatter(text)
                                if (frontmatter["name"]?.firstOrNull()?.matches(namePattern) != true) {
                                    add(
                                        Finding(
                                            severity,
                                            category,
                                            ctx.manifest.stringValue(category, "missingName")
                                                .takeIf { message -> message.isNotEmpty() }
                                                ?: "agent missing name: ${file.relativeTo(ctx.root)}",
                                        ),
                                    )
                                }
                                if (frontmatter["description"]?.firstOrNull()?.isNotBlank() != true) {
                                    add(
                                        Finding(
                                            severity,
                                            category,
                                            ctx.manifest.stringValue(category, "missingDescription")
                                                .takeIf { message -> message.isNotEmpty() }
                                                ?: "agent missing description: ${file.relativeTo(ctx.root)}",
                                        ),
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    private fun extractFrontmatter(content: String): Map<String, List<String>> =
        YamlFrontMatterVisitor()
            .also { visitor -> markdownParser.parse(content).accept(visitor) }
            .data
}
